Show the dataframe shape.

Standardize header names.

Which columns are numerical?

Which columns are categorical?

Check and deal with NaN values.

Datetime format- Extract the months from the dataset and store in a separate column. Then filter the data to show only the information for the first quarter , ie Jan Feb and March

BONUS: Put all the previously mentioned data transformations into a function.

In [19]:
import numpy as np

In [20]:
import pandas as pd

In [6]:
#mca = pd.read_csv("marketing_customer_analysis.csv")

In [21]:
dtf1 = pd.read_csv('file1.csv')
dtf2 = pd.read_csv('file2.csv')
dtf3 = pd.read_csv('file3.csv')

In [22]:
print(dtf1.shape)
print(dtf2.shape)
print(dtf3.shape)

(4008, 11)
(996, 11)
(7070, 11)


Standardize header names

In [23]:
# check the problem

dtf1.columns.to_list()

['Customer',
 'ST',
 'GENDER',
 'Education',
 'Customer Lifetime Value',
 'Income',
 'Monthly Premium Auto',
 'Number of Open Complaints',
 'Policy Type',
 'Vehicle Class',
 'Total Claim Amount']

Problems identified: clear whitespace and replace with _, change inconsistency in cases to all lowercase

In [24]:
def standardize_col(col):
    return col.lower().replace(" ", "_")

In [25]:
# remove the whitespace from the string Total Claim Amount
standardize_col("Total Claim Amount")

'total_claim_amount'

In [27]:
for col in dtf1.columns.to_list():
    print(standardize_col(col))

customer
st
gender
education
customer_lifetime_value
income
monthly_premium_auto
number_of_open_complaints
policy_type
vehicle_class
total_claim_amount


In [28]:
# now we create a list were we store all these standardized
# columns in
stdzd_cols = []
for col in dtf1.columns.to_list():
    stdzd_cols.append(standardize_col(col))

In [ ]:
# overwrite these columns with the original columns

In [30]:
dtf1.columns = stdzd_cols

In [31]:
# checking the changes
dtf1.columns

Index(['customer', 'st', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

perfom this job on two other dataframes.
So write a function that receives a dataframe and standardizes the columns

In [32]:
def standardize_dtf_colnames(dtf):
    #this is just copy pasted from above
    stdzd_cols = []
    for col in dtf.columns.to_list():
        # making use of the standardize_col function from above
        stdzd_cols.append(standardize_col(col))
    # in the end, overwrite columns with standardized ones
    dtf.columns = stdzd_cols

In [34]:

# standardize df2
standardize_dtf_colnames(dtf2)

# standardize df3
standardize_dtf_colnames(dtf3)

In [35]:
dtf2.columns

Index(['customer', 'st', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'total_claim_amount', 'policy_type', 'vehicle_class'],
      dtype='object')

In [36]:
dtf3.columns

Index(['customer', 'state', 'customer_lifetime_value', 'education', 'gender',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'total_claim_amount', 'vehicle_class'],
      dtype='object')

Why do we standardize?
it's more convenient working with a dataframe for which you don't have to remember exactly which column name is capital / lower case letters
sometimes you see that columns are not refered to as df3['vehicle_class'] but df3.vehicle_class. The latter would not be possible if you dont have the _ between the words

## Concatenate dataframes¶
How should we concatenate the dataframes? Stick them together horizontally (column wise) or vertically (row wise). Let's look at them first

In [37]:
dtf1.head()

customer          st gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [38]:
dtf2.head()

customer          st gender education customer_lifetime_value  income  \
0  GS98873     Arizona      F  Bachelor              323912.47%   16061   
1  CW49887  California      F    Master              462680.11%   79487   
2  MY31220  California      F   College              899704.02%   54230   
3  UH35128      Oregon      F   College             2580706.30%   71210   
4  WH52799     Arizona      F   College              380812.21%   94903   

   monthly_premium_auto number_of_open_complaints  total_claim_amount  \
0                    88                    1/0/00               633.6   
1                   114                    1/0/00               547.2   
2                   112                    1/0/00               537.6   
3                   214                    1/1/00              1027.2   
4                    94                    1/0/00               451.2   

      policy_type  vehicle_class  
0   Personal Auto  Four-Door Car  
1    Special Auto            SUV  
2   Personal Auto   Two-Door Car  
3   Personal Auto     Luxury Car  
4  Corporate Auto   Two-Door Car

In [39]:
dtf3.head()

customer       state  customer_lifetime_value             education gender  \
0  SA25987  Washington              3479.137523  High School or Below      M   
1  TB86706     Arizona              2502.637401                Master      M   
2  ZL73902      Nevada              3265.156348              Bachelor      F   
3  KX23516  California              4455.843406  High School or Below      F   
4  FN77294  California              7704.958480  High School or Below      M   

   income  monthly_premium_auto  number_of_open_complaints    policy_type  \
0       0                   104                          0  Personal Auto   
1       0                    66                          0  Personal Auto   
2   25820                    82                          0  Personal Auto   
3       0                   121                          0  Personal Auto   
4   30366                   101                          2  Personal Auto   

   total_claim_amount  vehicle_class  
0          499.200000   Two-Door Car  
1            3.468912   Two-Door Car  
2          393.600000  Four-Door Car  
3          699.615192            SUV  
4          484.800000            SUV

column wise it is! We don't need to think about ordering the column names of the individual data frames. Pandas takes care of that when concatenating

In [40]:
# we concatenate rowwise. We don't need to change "axis" parameter,
# because axis=0 (rowwise) is the default
dtf_all = pd.concat([dtf1, dtf2, dtf3],
                  # axis=0,
                  )

In [41]:
dtf_all # this is merging all the dataframe to 1 dataframe

customer          st gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
7065  LA72316         NaN      M              Bachelor   
7066  PK87824         NaN      F               College   
7067  TD14365         NaN      M              Bachelor   
7068  UP19263         NaN      M               College   
7069  Y167826         NaN      M               College   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
7065             23405.98798  71941.0                  73.0   
7066             3096.511217  21604.0                  79.0   
7067             8163.890428      0.0                  85.0   
7068             7524.442436  21941.0                  96.0   
7069             2611.836866      0.0                  77.0   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
7065                         0   Personal Auto  Four-Door Car   
7066                         0  Corporate Auto  Four-Door Car   
7067                         3  Corporate Auto  Four-Door Car   
7068                         0   Personal Auto  Four-Door Car   
7069                         0  Corporate Auto   Two-Door Car   

      total_claim_amount       state  
0               2.704934         NaN  
1            1131.464935         NaN  
2             566.472247         NaN  
3             529.881344         NaN  
4              17.269323         NaN  
...                  ...         ...  
7065          198.234764  California  
7066          379.200000  California  
7067          790.784983  California  
7068          691.200000  California  
7069          369.600000  California  

[12074 rows x 12 columns]


Now, we need to reset the index, because the index numbers are still from the original dataframe (check e.g. that we have 12074 rows, but our index ends at 7069)

In [42]:
dtf_all = dtf_all.reset_index(drop=True)
# or do (same effect)
# df_all.reset_index(drop=True, inplace=True)

In [43]:
dtf_all

customer          st gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316         NaN      M              Bachelor   
12070  PK87824         NaN      F               College   
12071  TD14365         NaN      M              Bachelor   
12072  UP19263         NaN      M               College   
12073  Y167826         NaN      M               College   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                  697953.59%      0.0                  94.0   
2                 1288743.17%  48767.0                 108.0   
3                  764586.18%      0.0                 106.0   
4                  536307.65%  36357.0                  68.0   
...                       ...      ...                   ...   
12069             23405.98798  71941.0                  73.0   
12070             3096.511217  21604.0                  79.0   
12071             8163.890428      0.0                  85.0   
12072             7524.442436  21941.0                  96.0   
12073             2611.836866      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                        1/0/00   Personal Auto  Four-Door Car   
1                        1/0/00   Personal Auto  Four-Door Car   
2                        1/0/00   Personal Auto   Two-Door Car   
3                        1/0/00  Corporate Auto            SUV   
4                        1/0/00   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
12069                         0   Personal Auto  Four-Door Car   
12070                         0  Corporate Auto  Four-Door Car   
12071                         3  Corporate Auto  Four-Door Car   
12072                         0   Personal Auto  Four-Door Car   
12073                         0  Corporate Auto   Two-Door Car   

       total_claim_amount       state  
0                2.704934         NaN  
1             1131.464935         NaN  
2              566.472247         NaN  
3              529.881344         NaN  
4               17.269323         NaN  
...                   ...         ...  
12069          198.234764  California  
12070          379.200000  California  
12071          790.784983  California  
12072          691.200000  California  
12073          369.600000  California  

[12074 rows x 12 columns]

Let's insert a data cleaning step here already and delete all the rows that contain only NaN values. Because what's the point in having them? 

In [44]:
# creates you a boolean mask for every row that has NaN values only! check also .any() etc...
dtf_all.isna().all(axis=1)

0        False
1        False
2        False
3        False
4        False
         ...  
12069    False
12070    False
12071    False
12072    False
12073    False
Length: 12074, dtype: bool

filter the rows by applying this "sieve" to the dataframe. We're using ~ here because we want to negate / invert the boolean "sieve" and keep the rows where it says False

In [45]:

dtf_all = dtf_all[~dtf_all.isna().all(axis=1)]

In [46]:

dtf_all = dtf_all.reset_index(drop=True)

## drop duplicate rows

In [47]:
dtf_all = dtf_all.drop_duplicates()

## filter negative incomes
because we regard those as data errors

In [49]:
dtf_all = dtf_all[dtf_all['income'] >= 0]

## Which columns are numerical?

see documentation https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.select_dtypes.html

In [51]:
dtf_all.select_dtypes(np.number)

income  monthly_premium_auto  total_claim_amount
0         0.0                1000.0            2.704934
1         0.0                  94.0         1131.464935
2     48767.0                 108.0          566.472247
3         0.0                 106.0          529.881344
4     36357.0                  68.0           17.269323
...       ...                   ...                 ...
9132  71941.0                  73.0          198.234764
9133  21604.0                  79.0          379.200000
9134      0.0                  85.0          790.784983
9135  21941.0                  96.0          691.200000
9136      0.0                  77.0          369.600000

[9134 rows x 3 columns]

In [52]:
# or more convenient
dtf_nums = dtf_all._get_numeric_data()

## Which are categorical?
Short answer? All others 

In [53]:
# drop all numerical columns
dtf_cats = dtf_all.drop(columns=dtf_nums.columns) #cat is for categororical

In [54]:
dtf_cats

customer          st gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
9132  LA72316         NaN      M              Bachelor   
9133  PK87824         NaN      F               College   
9134  TD14365         NaN      M              Bachelor   
9135  UP19263         NaN      M               College   
9136  Y167826         NaN      M               College   

     customer_lifetime_value number_of_open_complaints     policy_type  \
0                        NaN                    1/0/00   Personal Auto   
1                 697953.59%                    1/0/00   Personal Auto   
2                1288743.17%                    1/0/00   Personal Auto   
3                 764586.18%                    1/0/00  Corporate Auto   
4                 536307.65%                    1/0/00   Personal Auto   
...                      ...                       ...             ...   
9132             23405.98798                         0   Personal Auto   
9133             3096.511217                         0  Corporate Auto   
9134             8163.890428                         3  Corporate Auto   
9135             7524.442436                         0   Personal Auto   
9136             2611.836866                         0  Corporate Auto   

      vehicle_class       state  
0     Four-Door Car         NaN  
1     Four-Door Car         NaN  
2      Two-Door Car         NaN  
3               SUV         NaN  
4     Four-Door Car         NaN  
...             ...         ...  
9132  Four-Door Car  California  
9133  Four-Door Car  California  
9134  Four-Door Car  California  
9135  Four-Door Car  California  
9136   Two-Door Car  California  

[9134 rows x 9 columns]

After the data cleaning... we are left with 2 data, which are categorical and numerical data.

# df_nums.
# df_cats

# checking whats dirty

In [55]:
# check for number of null values in every column
dtf_cats.isna().sum()

customer                        0
st                           7070
gender                        122
education                       0
customer_lifetime_value         7
number_of_open_complaints       0
policy_type                     0
vehicle_class                   0
state                        2064
dtype: int64

In [56]:

dtf_cats.isna().sum()/len(dtf_cats)

customer                     0.000000
st                           0.774031
gender                       0.013357
education                    0.000000
customer_lifetime_value      0.000766
number_of_open_complaints    0.000000
policy_type                  0.000000
vehicle_class                0.000000
state                        0.225969
dtype: float64

In [ ]:
# Perform thesame for numerical

In [57]:
dtf_nums.isna().sum()/len(dtf_nums)

income                  0.0
monthly_premium_auto    0.0
total_claim_amount      0.0
dtype: float64

In [ ]:
# NOTE: Numerics are clean of null values..... St!

In [58]:
dtf_cats['st'].value_counts()

Oregon        623
California    486
Arizona       327
Nevada        223
Washington    181
Cali          120
AZ             74
WA             30
Name: st, dtype: int64

In [ ]:

# NOTE: This checks up the error in..... 'gender'

In [59]:
dtf_cats['gender'].value_counts()

F         4557
M         4368
Male        40
female      30
Femal       17
Name: gender, dtype: int64

In [ ]:

# Now let's clean all the entries that are not F or M and turn it into either F or M

In [60]:
# a task like this can only be solved manually
# a good approach here is a dictionary, because, what do we want to do? Translating!

trans_dct = {'Male': 'M',
             'female':'F',
             'Femal':'F',
             'F':'F',
             'M':'M',
            }

# now, we can write ourselve a function, that does the translate
# and use df_cats['gender'].apply(my_function)
# but we can also make use of the map function, which is way faster (although speed is not 
# important at this point yet)!
dtf_cats['gender'] = dtf_cats['gender'].map(trans_dct)

In [61]:

dtf_cats['gender'].value_counts()

F    4604
M    4408
Name: gender, dtype: int64

In [ ]:
# checks error in 'education'

In [62]:

dtf_cats['education'].value_counts()

Bachelor                2718
College                 2681
High School or Below    2616
Master                   751
Doctor                   344
Bachelors                 24
Name: education, dtype: int64

In [63]:
# this checks the 'customer_lifetime_value'

In [64]:
dtf_cats['customer_lifetime_value'].value_counts()

13727.79972    6
6366.225775    6
3954.344534    6
4349.30875     6
5568.947534    6
              ..
5137.349322    1
5133.414764    1
5132.684537    1
747831.37%     1
12564.06775    1
Name: customer_lifetime_value, Length: 8211, dtype: int64

In [65]:

def clean_clv(x):
    return float(str(x).replace('%', ''))*100

dtf_cats['customer_lifetime_value'] = dtf_cats['customer_lifetime_value'].apply(clean_clv)

In [ ]:
# this checks 'number of open complaints'

In [66]:

dtf_cats['number_of_open_complaints'].value_counts()

0         5629
1/0/00    1623
1          765
2          283
1/1/00     247
3          230
4          119
1/2/00      93
1/3/00      60
5           44
1/4/00      29
1/5/00      12
Name: number_of_open_complaints, dtype: int64

We see that we have either values of 0, 1, 2, 3, 4 or 5. But then also 1/0/00, 1/1/00, 1/2/00, 1/3/00, 1/4/00 and 1/5/00,
which contain only redundant info except the middle number.
So we create ourselves again a trans dict and roll like above:

In [67]:
trans_dct = {0:0,
             1:1,
             2:2,
             3:3,
             4:4,
             5:5,
             '1/0/00' :0, 
             '1/1/00' :1, 
             '1/2/00' :2, 
             '1/3/00' :3, 
             '1/4/00' :4, 
             '1/5/00' :5, 
            }

dtf_cats['number_of_open_complaints'] = dtf_cats['number_of_open_complaints'].map(trans_dct)

In [69]:
dtf_cats['number_of_open_complaints'].value_counts()

0    7252
1    1012
2     376
3     290
4     148
5      56
Name: number_of_open_complaints, dtype: int64

Clean Data!!